### DISCUSSION TITLE
dev not picking up comments in pyproject.toml or reload not working

### DISCUSSION THREAD

**Discussion by j1philli** [pkgxdev/pkgx#768] [Help] [answered]
HI!
Just switched from tea to pkgx and went to get dev setup.
![CleanShot 2023-10-02 at 22 31 46](https://github.com/pkgxdev/pkgx/assets/3744255/ba40d899-7bda-4981-9679-f65bb82891ac)
Because of issues with python 3.12 and my project I want to have it use 3.11.4 so I put it in the pyproject.toml
![CleanShot 2023-10-02 at 22 33 12](https://github.com/pkgxdev/pkgx/assets/3744255/39ce3f6d-5ff2-4771-8af5-4f7cc5293f77)
But alas I still have 3.12 in my dev environment
![CleanShot 2023-10-02 at 22 33 39](https://github.com/pkgxdev/pkgx/assets/3744255/decb3962-fe7d-4774-9be0-e1f210d23191)
Any guidance would be much appreciated!

**Comment by mxcl** [comment]
Does `cd .. && cd -` work? `cd .` should work but just checking.

@jhheider I guess the “don't use python 3.12 as a dep yet” code needs enacting sooner or we gotta go through the pantry and change python reqs.

**Reply by jhheider** [reply]
Explicit deps is better than mysterious suppression. I'll get those changed today.

**Reply by mxcl** [reply]
I probs will add the “don’t jump to x.y.0 releases for things like python” as a dep functionality at some point. It's good for robustness. We can output an informational message when we do it that they can override with a `+python@3.12` addition.

However the liklihood that everything that uses python will update in a _knowable_ timeframe is slim, so not sure it buys us a lot.

The problem with editing all the pacakges to dep on 3.11 is they are likely to remain that way forever. But yes, it's a good choice for *today*.

**Reply by jhheider** [reply]
Initial state:
```shell
$ rg -NI --trim python.org:\  projects | sort -u
python.org: ">=3.7"
python.org: ">=3.7<3.12"
python.org: '*'
python.org: '*' #FIXME actually: ^3.7
python.org: '>=3.10<3.12'
python.org: '>=3.11'
python.org: '>=3.7 <3.11'
python.org: '>=3.7'
python.org: '>=3.8'
python.org: '^3.11'
python.org: '^3.7'
python.org: '~3.11'
python.org: 3
python.org: 3 #TODO ^3.5
python.org: 3.11
python.org: =3.11
python.org: ^3
python.org: ^3.1
python.org: ^3.10
python.org: ^3.11
python.org: ^3.7
python.org: ^3.8
python.org: ~3.10
python.org: ~3.11
```

The fix:
```shell
$ rg -l python.org:\  projects/ -0 |\
  xargs -0 sed -i.bak \
    -e "s/python.org: \">=3.7\"/python.org: '>=3.7<3.12'/" \
    -e "s/python.org: '\*'/python.org: '<3.12'/" \
    -e "s/python.org: '>=3.11'/python.org: ~3.11/" \
    -e "s/python.org: '>=3.7'/python.org: '>=3.7<3.12'/" \
    -e "s/python.org: '>=3.8'/python.org: '>=3.8<3.12'/" \
    -e "s/python.org: '\^3.11'/python.org: ~3.11/" \
    -e "s/python.org: '\^3.7'/python.org: '>=3.7<3.12'/" \
    -e "s/python.org: '\~3.11'/python.org: ~3.11/" \
    -e "s/python.org: 3$/python.org: '>=3<3.12'/" \
    -e "s/python.org: 3.11/python.org: ~3.11/" \
    -e "s/python.org: =3.11/python.org: ~3.11/" \
    -e "s/python.org: \^3$/python.org: '>=3<3.12'/" \
    -e "s/python.org: \^3.1$/python.org: '>=3.1<3.12'/" \
    -e "s/python.org: \^3.10/python.org: '>=3.10<3.12'/" \
    -e "s/python.org: \^3.11/python.org: ~3.11/" \
    -e "s/python.org: \^3.7/python.org: '>=3.7<3.12'/" \
    -e "s/python.org: \^3.8/python.org: '>=3.8<3.12'/" && \
  find projects -name '*.bak' -exec rm {} \;
```

End state:
```shell
$ rg -NI --trim python.org:\  projects | sort -u
python.org: ">=3.7<3.12"
python.org: '*' #FIXME actually: ^3.7    <~~~~ this is pip.pypa.io
python.org: '<3.12'
python.org: '>=3.10<3.12'
python.org: '>=3.1<3.12'
python.org: '>=3.5<3.12'
python.org: '>=3.7<3.11'
python.org: '>=3.7<3.12'
python.org: '>=3.8<3.12'
python.org: '>=3<3.12'
python.org: '>=3<3.12' #FIXME rq’d by gnome.org/gobject-introspection but should be added by tea-env
python.org: ~3.10
python.org: ~3.11
```

**Reply by jhheider** [reply]
> don’t jump to x.y.0 releases for things like python

python is certainly one. llvm likely another. probably the optimal way is yaml key for `minimum-implicit-patch`. It shouldn't make the dep hydration algorithm too awful, since you can just exclude those below the threshold.

**Reply by j1philli** [reply]
So it looks like it did?
![CleanShot 2023-10-03 at 10 23 44](https://github.com/pkgxdev/pkgx/assets/3744255/9ab87e65-9dda-49f0-b82c-268f0bec18a6)
However it does not downgrade my python version based on my pyproject.toml

**Reply by j1philli** [reply]
@jhheider are these changes in pantry working as intended? pkgx seems to still be installing 3.12 for this [package](https://github.com/pkgxdev/pantry/blob/fe4553718c175c234572c0eed70dabdbd3369e47/projects/python-poetry.org/package.yml#L9)
talking about it here: https://github.com/orgs/pkgxdev/discussions/791

**Reply by jhheider** [reply]
That change is recent. Does it appear in your local pantry under ~/.pkgx/pkgx.sh/var/pantry/projects? If not, `pkgx --sync` should grab it.

**Reply by j1philli** [reply]
I do not see ~/.pkgx/pkgx.sh/var/ even after ```pkgx --sync```.
However it does now pull the correct version
<img width="641" alt="image" src="https://github.com/pkgxdev/pkgx/assets/3744255/4356da66-32a1-4b72-b6d4-bc85008f6150">

**Reply by jhheider** [reply]
Ah, right, we're under $XDG_DATA_HOME/pkgx/pantry if set, and ~/Library/Application\ Support/pkgx/pantry on darwin or ~/.local/share/pkgx/pantry on linux otherwise.

**Reply by j1philli** [reply]
ah very cool thanks! So the pantry can be fetched independently of pkgx? I assumed I had to wait for a pkgx update.
Does the command do anything else?
I don't think the --sync command is in docs. I could add it

**Comment by j1philli** [comment, marked as answer]
Created https://github.com/pkgxdev/pkgx/issues/774 now that we know the refresh does not pick up the comments or python declaration
